Fix --bind-address problem identified by issue 197

yu2003w opened the pull request:
Since kafka 0.9.0, option "listeners" has been introduced. 
However, it doesn't compatible with '--bind-address" which is an parameter of kafka mesos scheduler.
To fix the problem, if "--bind-address" is specified, "host" of "listeners" should be set as "offer.getHostname".
This fix was tested in development environment.

yu2003w committed:
Fix --bind-address problem identified by issue 197

Since kafka 0.9.0, option "listeners" has been introduced. However, it doesn't add related support '--bind-address" which is an parameter of kafka mesos scheduler.
To fix this problem, if "--bind-address" is specified, "host" of "listeners" should be set as "offer.getHostname".
This fix was tested in development environment.

steveniemitz commented on src/scala/ly/stealth/mesos/kafka/Scheduler.scala:
```diff
@@ -87,8 +87,14 @@ object Scheduler extends org.apache.mesos.Scheduler {
-        defaults += ("listeners" -> s"PLAINTEXT://:${reservation.port}")
+      if (HttpServer.kafkaVersion.compareTo(new Version("0.9")) >= 0) {
+        if (broker.bindAddress != null) {
+          defaults += ("listeners" -> s"PLAINTEXT://${offer.getHostname}:${reservation.port}")
```
I think you should be using the result of bindAddress.resolve() here for the listener address rather than the offer, since the bindAddress could be a different interface than the slave process is configured to advertise.

Once you change it, you'll have to then move this logic into the BrokerServer code so it runs on the slave itself and can correctly resolve the address.  This logic currently exists in Broker.options(...) to set the host.name property correctly on the slave.

yu2003w replied:
@steveniemitz 
I tried to add broker as below and then started,
./kafka-mesos.sh broker add 0 --heap 256 --mem 512 --cpu 0.2 --bind-address if:eth1

Code mentioned above is in Scheduler.newTask(). This function is invoked in driver after Scheduler.acceptOffer() is called. So here, bindAddress.resolve() will return IP address of driver rather than that of mesos agent which will launch broker. 

The call stack for "--bind-address" scenario is almost like below,
 at ly.stealth.mesos.kafka.Util$BindAddress.resolveInterfaceAddress(Util.scala:111)
        at ly.stealth.mesos.kafka.Util$BindAddress.resolve(Util.scala:92)
        at ly.stealth.mesos.kafka.Scheduler$.taskData$1(Scheduler.scala:92)
	at ly.stealth.mesos.kafka.Scheduler$.newTask(Scheduler.scala:113)
	at ly.stealth.mesos.kafka.Scheduler$.launchTask(Scheduler.scala:318)
	at ly.stealth.mesos.kafka.Scheduler$$anonfun$acceptOffer$2.apply(Scheduler.scala:240)
	at ly.stealth.mesos.kafka.Scheduler$$anonfun$acceptOffer$2.apply(Scheduler.scala:236)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at ly.stealth.mesos.kafka.Scheduler$.acceptOffer(Scheduler.scala:236)
	at ly.stealth.mesos.kafka.Scheduler$$anonfun$syncBrokers$1.apply(Scheduler.scala:203)
	at ly.stealth.mesos.kafka.Scheduler$$anonfun$syncBrokers$1.apply(Scheduler.scala:202)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at ly.stealth.mesos.kafka.Scheduler$.syncBrokers(Scheduler.scala:202)
	at ly.stealth.mesos.kafka.Scheduler$.resourceOffers(Scheduler.scala:139)

After offer is accepted, executor will resolve address based on "--bind-address" on mesos agent node. If everything goes well, broker will be started.
Back to the test case I ran, when the agent did not have an interface named "eth1", it failed with following exceptions,
java.lang.IllegalStateException: Failed to resolve if:eth1
	at ly.stealth.mesos.kafka.Util$BindAddress.resolveInterfaceAddress(Util.scala:111)
	at ly.stealth.mesos.kafka.Util$BindAddress.resolve(Util.scala:92)
	at ly.stealth.mesos.kafka.Broker.options(Broker.scala:71)
	at ly.stealth.mesos.kafka.KafkaServer.start(BrokerServer.scala:50)
	at ly.stealth.mesos.kafka.Executor$.ly$stealth$mesos$kafka$Executor$$runBroker0$1(Executor.scala:89)
	at ly.stealth.mesos.kafka.Executor$$anon$1.run(Executor.scala:112)

Actually scheduler only could get hostname of slave advertised. So here, "listeners" host could be set as offer.getHostname.

Any other comments?

steveniemitz replied:
The stack trace you posted is the existing call to `result.put("host.name", bindAddress.resolve())`.  My point was you need to add the code that sets the listener address in the same place as that existing call so it runs on the executor and not the scheduler.

yu2003w replied:
@steveniemitz 
Thanks for your comments. I misunderstood something in your first post.
I discarded changes for scheduler.scala and added logic in Broker.options() to set "listeners" with result of bindAddress.resolve(). I tested the changes in my environment.
Any other comments please let me know.

yu2003w committed:
Replace Broker option "listeners" with correct IP address when "--bind-address" is specified.
This commit is to fix "--bind-address" issue since Kafka 0.9.
The detailed description is in issue_197.

yu2003w committed:
No need to change Scheduler.scala to fix the problem. Discard previous changes.

steveniemitz commented on src/scala/ly/stealth/mesos/kafka/Broker.scala:
```diff
@@ -67,8 +67,11 @@ class Broker(_id: String = "0") {
-    if (bindAddress != null)
+    if (bindAddress != null) {
       result.put("host.name", bindAddress.resolve())
+      if (result.get("listeneres") != null)
```
typeo in "listeneres".  Are there any unit tests around this functionality?  If not could you add them to verify this works.

Also, the hardcoded strings here are less than ideal, I'd save the bindAddress.resolve() result in a local variable and use it in both places instead of looking up the value just added.

yu2003w replied:
@steveniemitz 
Really sorry it's typo. I updated code again with unit tests added for "listeners" and "bindAddress".
Thanks for comments.

yu2003w committed:
Added unit test cases and fixed typos.

steveniemitz commented:
Cool, I'd help merge this if I could but unfortunately the only people with write access to this repo disappeared. :(

yu2003w commented:
@steveniemitz 
I talked with some guys from mesosphere. It seemed that the original author did not want to maintain this project anymore. If so, it's really a pity.
Thanks very much for your review.

steveniemitz commented:
This was merged in ed0ff0732990c2fc4a7f3ef192b15a34ca2f12b3
